File: 59_SpiralMatrixII.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        matrix = [[None]*n for _ in range(n)]
        row = col = count = 0
        size = n-1
        
        for num in range(1, n**2 + 1):
            if matrix[row][col]:
                row += 1
                col += 1
                count += 1
                size -= 1

            matrix[row][col] = num
            
            if row == count and col < size: col += 1
            elif row < size and col == size : row += 1
            elif row == size and col > count : col -= 1
            elif row > count and col == count: row -= 1
                
        return matrix
        

File: swift/59_SpiralMatrixII.swift
class Solution {
    func generateMatrix(_ n: Int) -> [[Int]] {
        var spiralMatrix = [[Int]](repeating: [Int](repeating: 0, count: n), count: n)
        var i = 0, j = 0, layer = 0, sz = n - 1
        
        for x in 1...n*n {
            if spiralMatrix[i][j] != 0 {
                layer += 1
                sz -= 1
                i = layer
                j = layer
            }
            
            spiralMatrix[i][j] = x
            
            if i == layer && j < sz { j += 1 }
            else if j == sz && i < sz { i += 1 }
            else if i == sz && j > layer { j -= 1 }
            else { i -= 1 }
        }
        return spiralMatrix
    }
}
